File: problems/arrays_and_hashing/cpp/169_majority_element.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int majorityElement(vector<int> &nums) {
    int candidate = nums[0], vote = 1;
    for (size_t i = 1; i < nums.size(); i++) {
      if (vote == 0) {
        candidate = nums[i];
        vote++;
      } else if (candidate != nums[i]) {
        vote--;
      } else {
        vote++;
      }
    }
    return candidate;
  }
};

// OR
// Time Complexity: O(n) - single pass through the array
// Space Complexity: O(1) - only uses two variables

class Solution {
public:
  int majorityElement(vector<int> &nums) {
    int candidate = 0;
    int count = 0;

    for (int num : nums) {
      if (count == 0) {
        candidate = num;
      }
      count += (num == candidate) ? 1 : -1;
    }

    return candidate;
  }
};

File: problems/arrays_and_hashing/rust/169_majority_element.rs
// Time Complexity: O(n) - single pass through the array
// Space Complexity: O(1) - only uses two variables

impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut candidate = 0;
        let mut count = 0;

        for num in nums {
            if count == 0 {
                candidate = num;
            }
            count += if num == candidate { 1 } else { -1 };
        }

        candidate
    }
}
